tools/testbench/testbench.c

```diff
@@ -353,13 +353,14 @@ static int parse_input_args(int argc, char **argv, struct testbench_prm *tp)
 			break;
 
 		/* print usage */
-		default:
-			fprintf(stderr, "unknown option %c\n", option);
-			ret = -EINVAL;
-			/* fallthrough */
 		case 'h':
 			print_usage(argv[0]);
 			exit(EXIT_SUCCESS);
+
+		default:
+			fprintf(stderr, "unknown option %c\n", option);
+			print_usage(argv[0]);
+			ret = -EINVAL;
```

Comment:
You're getting away with this problem with a refactoring this time around but it's "selfish"! How about other statements elsewhere in the code?

Could you try to find something portable across both gcc and clang?
https://stackoverflow.com/questions/45129741/gcc-7-wimplicit-fallthrough-warnings-and-portable-way-to-clear-them

Reply:
I'll try if there's something that works for all compilers with help of that link (yesterday I didn't find).

Reply:
Can you please try something like this? It supposedly works with both gcc and clang.

```C
#if defined __has_cpp_attribute && __has_cpp_attribute(fallthrough)
#  define SOF_FALLTHRU __attribute__((__fallthrough__)
#elif <FUTURE_UNUSUAL_COMPILER>
#  define SOF_FALLTHRU <something unusual>
#else
#  warning unknown compiler, undefined SOF_FALLTHRU
#endif

```

Reply:
https://github.com/zephyrproject-rtos/zephyr/tree/main/include/zephyr/toolchain uses `__attribute__((__fallthrough__)` for both gcc and clang.

Reply:
It's used with #define like this:

#ifndef __fallthrough
#if __GNUC__ >= 7
#define __fallthrough        __attribute__((fallthrough))
#else
#define __fallthrough
#endif	/* __GNUC__ >= 7 */
#endif

Reply:
I still feel my code re-order is simplest way to address this. The fall-through is unnecessary trickery for a very simple thing.

Reply:
Also if you look the original code, there's a bug that I'm fixing the same time:

```
		default:
			fprintf(stderr, "unknown option %c\n", option);
			ret = -EINVAL;
			/* fallthrough */
		case 'h':
			print_usage(argv[0]);
			exit(EXIT_SUCCESS);
		}

		if (ret < 0)
			return ret;

```

The `-EINVAL `is not returned because of `exit(EXIT_SUCCESS)`. Fall-through is not correct here.

Reply:
> unnecessary trickery for a very simple thing

I never said your change was wrong, I said it was "selfish" and I stand by that. This type of issue is painful to test because you need multiple toolchains and many people use only one. But you already have the environment, skills and a good test case. 

That "trickery" would really not be that complicated (try `git grep META` for some _real_ trickery!) and it would be only in one or two .h files; just one very simple SOF_FALLTHRU keyword everywhere else.


> `#if __GNUC__ >= 7` 

We don't need stuff like that for `testbench.c`, so it could be very simple for now.

Reply:
> Also if you look the original code, there's a bug that I'm fixing the same time: [...]
> Fall-through is not correct here.

The current commit message still refers to fall through only.

Reply:
@marc-hb I just updated the commit text. Is it OK now?

Reply:
Sure. Just for the record I never formally "requested" fallthrough changes :-)